packages/go_router/lib/src/route.dart

```diff
@@ -446,22 +473,201 @@ class ShellRoute extends RouteBase {
+///         ),
+///       ],
+///     ),
+///   ],
+/// );
+/// ```
+///
+/// See [Stateful Nested Navigation](https://github.com/flutter/packages/blob/main/packages/go_router/example/lib/stateful_nested_navigation.dart)
+/// for a complete runnable example.
+class PartitionedShellRoute extends ShellRouteBase {
+  /// Constructs a [PartitionedShellRoute] from a list of routes, each
+  /// representing the root of a navigation hierarchy. A separate
+  /// [Navigator] will be created for each of the root routes, using the
+  /// navigator key at the corresponding index in [navigatorKeys].
+  PartitionedShellRoute({
+    required List<RouteBase> routes,
+    required this.navigatorKeys,
+    super.builder,
+    super.pageBuilder,
+  })  : assert(routes.isNotEmpty),
+        assert(navigatorKeys.length == routes.length),
+        assert(Set<GlobalKey<NavigatorState>>.from(navigatorKeys).length ==
+            routes.length),
+        assert(pageBuilder != null || builder != null,
+            'builder or pageBuilder must be provided'),
+        super._(routes: routes) {
+    for (int i = 0; i < routes.length; ++i) {
+      final RouteBase route = routes[i];
+      if (route is GoRoute) {
+        assert(route.parentNavigatorKey == null ||
+            route.parentNavigatorKey == navigatorKeys[i]);
+      }
+    }
+  }
+
+  /// Constructs a [PartitionedShellRoute] that places its navigators in an
+  /// [IndexedStack], managed by a [StackedNavigationShell].
+  ///
+  /// Each route in the `routes` parameter must correspond to a
+  /// [StackedNavigationItem], specified in the `stackItems` parameter.
+  /// The stacked navigation shell can be implemented by specifying a
+  /// `scaffoldBuilder`, to build a widget that wraps the indexed stack.
+  factory PartitionedShellRoute.stackedNavigationShell({
+    required List<RouteBase> routes,
+    required List<StackedNavigationItem> stackItems,
+    StackedNavigationScaffoldBuilder? scaffoldBuilder,
+    StackedNavigationTransitionBuilder? transitionBuilder,
+    Duration? transitionDuration,
+  }) {
+    return PartitionedShellRoute(
+        routes: routes,
+        navigatorKeys: stackItems
+            .map((StackedNavigationItem e) => e.navigatorKey)
+            .toList(),
+        builder: (BuildContext context, GoRouterState state, Widget child) {
+          assert(child is Navigator);
+          return StackedNavigationShell(
+            currentNavigator: child as Navigator,
+            currentRouterState: state,
+            stackItems: stackItems,
+            scaffoldBuilder: scaffoldBuilder,
+            transitionBuilder: transitionBuilder,
+            transitionDuration: transitionDuration,
+          );
+        });
+  }
+
+  /// The navigator keys of the navigators created by this route.
+  final List<GlobalKey<NavigatorState>> navigatorKeys;
+
+  @override
+  GlobalKey<NavigatorState> navigatorKeyForChildRoute(RouteBase route) {
```

Comment:
it looks like the navigatorKeys/StackedNavigationItem needs to correspond to the routes in the same order otherwise it will either crash or have unexpected behavior. If that is the case, is it possible to make it so the StackedNavigationItem take a RouteBase object and make stackedNavigationShell takes a list of StackedNavigationItem as its subroutes?

instead of
```dart
PartitionedShellRoute.stackedNavigationShell(
  stackItems: [stackItems1, stackItems2, stackItems3]
  routes: [route1, route2, route3],
)
```
let it be something like this
```dart
PartitionedShellRoute.stackedNavigationShell(
  routes: [stackItems1(route: route1), stackItems2(route: route2), stackItems3(route: route3)],
)
```

Reply:
This needs some refactoring of the builder code an routebase class though

Reply:
Well, perhaps not crash, this was meant more as an assert as the idea was that the method should never be called with a child route that didn't exist. But this method is changed now to be able to return null, so it's no longer an issue. 

Regarding the constructors, there were asserts in place to prevents arrays of different lengths to be passed etc, so this should have been solid. However, this is anyway removed in the latest commits.